src/Service/ModuleVersionFactory.php

```diff
@@ -23,41 +26,75 @@ public function setModuleVersions(Installation $installation, object $installedM
-                $module->setName($name);
-                if (isset($installed->display_name)) {
-                    $module->setDisplayName($installed->display_name);
+            $module = $this->getModule($name, $installed->package);
+
+            if (isset($installed->display_name)) {
+                $module->setDisplayName($installed->display_name);
+            }
+            $module->setEnabled('Enabled' === $installed->status);
+
+            $moduleVersion = $this->getModuleVersion($module, $installed->version);
+            $moduleVersions->add($moduleVersion);
+        }
+
+        $installation->setModuleVersions($moduleVersions);
+
+        $this->entityManager->flush();
+        $this->createdModules = [];
+        $this->createdModuleVersions = [];
+    }
+
+    private function getModule(string $name, string $package): Module
+    {
+        $module = $this->moduleRepository->findOneBy([
+            'name' => $name,
+            'package' => $package,
+        ]);
+
+        if (null === $module) {
+            /** @var Module $createdModule */
+            foreach ($this->createdModules as $createdModule) {
+                if ($name === $createdModule->getName() && $package === $createdModule->getPackage()) {
+                    $module = $createdModule;
                 }
-                $module->setEnabled('Enabled' === $installed->status);
             }
+        }
+
+        if (null === $module) {
+            $module = new Module();
+            $this->entityManager->persist($module);
 
-            $moduleVersion = $this->moduleVersionRepository->findOneBy([
-                'module' => $module,
-                'version' => $installed->version,
-            ]);
+            $module->setName($name);
+            $module->setPackage($package);
+        }
 
-            if (null === $moduleVersion) {
-                $moduleVersion = new ModuleVersion();
-                $this->entityManager->persist($moduleVersion);
+        return $module;
+    }
 
-                $module->addModuleVersion($moduleVersion);
-                $installation->addModuleVersion($moduleVersion);
+    private function getModuleVersion(Module $module, ?string $version): ModuleVersion
+    {
+        $moduleVersion = $this->moduleVersionRepository->findOneBy([
+            'module' => $module,
+            'version' => $version,
+        ]);
+
+        if (null === $moduleVersion) {
+            /** @var ModuleVersion $createdModuleVersion */
+            foreach ($this->createdModuleVersions as $createdModuleVersion) {
+                if ($module->getId() === $createdModuleVersion->getModule()->getId() && $version === $createdModuleVersion->getVersion()) {
+                    $moduleVersion = $createdModuleVersion;
+                }
             }
+        }
 
-            $moduleVersion->setVersion($installed->version);
+        if (null === $moduleVersion) {
+            $moduleVersion = new ModuleVersion();
```

Comment:
Should this be added to `$createdModuleVersions`?

Reply:
Fixed